[1.0] Fix lookAt when head orientation is non-uniform

0b5vr opened the pull request:
### Description

TBD

### Points need review

- [ ] `VRMLookAt` の各メソッドについて、説明と得られる結果の整合性は取れていると思いますか？
- [ ] テスト、とりあえず今回の問題を解決しなきゃと思い、（もともとテストのなかった箇所に対して）すごいspecificで網羅的でないテストを書いてしまっていますが、一旦このままコミットしちゃっても良いですかね……？

0b5vr committed:
wip fix: Fix lookAt when head orientation is non-uniform

it still doesn't work properly

0b5vr commented:
あーしまった、Draftにするの忘れた

0b5vr commented on packages/three-vrm-core/src/lookAt/VRMLookAt.ts:
```diff
@@ -105,6 +105,11 @@ export class VRMLookAt {
+  /**
+   * World rotation of the head in its rest pose.
+   */
+  private _restHeadWorldQuaternion: THREE.Quaternion;
```
`protect` にするか迷ったが、一旦 `private` で。後でいつでも変えれる

ke456-png commented on packages/three-vrm-core/src/lookAt/VRMLookAt.ts:
```diff
@@ -206,12 +214,18 @@ export class VRMLookAt {
    * @param target A target `THREE.Vector3`
    */
   public getFaceFrontQuaternion(target: THREE.Quaternion): THREE.Quaternion {
-    if (this.faceFront.distanceToSquared(VEC3_POSITIVE_Z) < 0.01) {
-      return target.identity();
-    }
```
ここの処理が丸々無くなっていますが、大丈夫でしょうか？

0b5vr replied:
ここの処理は、headの回転が正規化されているという間違った前提で動いていたので、消えるのが正解です

0b5vr replied:
ここは、もともとはHeadが正規化されている前提で、「World Spaceの+Zを `FaceFront` 方向に向けるQuaternionを返す」関数でしたが、
今回これを、「World SpaceのHeadを `FaceFront` 方向に向けるQuaternionを返す」関数に変更しました。

即ち、 `WorldSpaceHead * FaceFrontQuaternion` を行うと、現在のheadのorientationでZ+が `faceFront` 方向を向くQuaternionを得られます。

実際にどのような用法が考えられるかは、 `VRMLookAtHelper` の用法を見ていただくのが早いと思います。

で、これは質問なのですが、

- このインタフェースは公開する価値があるものだと思いますか？
- 公開する場合、この `getFaceFrontQuaternion` というメソッド名は適切だと思いますか？
- また、以前の「World Spaceの+Zを `FaceFront` 方向に向けるQuaternionを返す」機能は別で必要だと思いますか？

ke456-png replied:
> このインタフェースは公開する価値があるものだと思いますか？

はい、使用したい場面は有り、公開する価値はあると思います。
Helperの様に表示したい場合はありそうです。

ke456-png replied:
> 公開する場合、この getFaceFrontQuaternion というメソッド名は適切だと思いますか？

はい、適切だと思います。
周辺の命名もget{Hoge}Quaternionという形になっているのでFaceFrontという単語が入るのは良いと思います。

変更する場合はFaceFrontとは別に命名を検討することになると思いますが、その必要性を感じません。

ke456-png replied:
> また、以前の「World Spaceの+Zを FaceFront 方向に向けるQuaternionを返す」機能は別で必要だと思いますか？

いいえ、必要ではないと思います。
しかし、自信はあまりないです。

ただLookAtの文脈でHeadを無視して+Zを FaceFront 方向に向けるシチュエーションが思いつきませんでした。

ke456-png reviewed (commented):
memo :
https://github.com/vrm-c/vrm-specification/blob/master/specification/VRMC_vrm-1.0-beta/lookAt.ja.md

ke456-png commented:
[Q] headとeyeの間に別のボーンが入ることはありますか？

その場合Applierで対応が必要そうですね

0b5vr committed:
wip fix: Fix lookAt when head orientation is non-uniform (2)

I believe the yaw and the pitch are working properly

it still doesn't work properly

0b5vr committed:
test: add a test, `VRMLookAt.lookAt`

0b5vr commented:
> [Q] headとeyeの間に別のボーンが入ることはありますか？
> 
> その場合Applierで対応が必要そうですね

あり得ます！

それでなくても現状、Applier側に、eyeのレスト回転が入っている場合に目が正常に動かない問題があります。

0b5vr committed:
test: slightly update look at test

0b5vr committed:
docs: fix error of docs

0b5vr committed:
fix: Fix lookAt when head orientation is non-uniform (3)

Now bone applier is working properly

0b5vr commented:
Readyのつもり。コメント追加していきます

0b5vr committed:
performance: revive faceFrontQuat calculation ditching branch

0b5vr committed:
refactor: lint-fix

0b5vr committed:
chore: three-vrm-core tests was not running using workspace root yarn test, fixed this

also removed outdated npm scripts

0b5vr committed:
test: add todo to lookAt tests

we should test the case when faceFront is (0, 0, -1) later

0b5vr commented:
`toBeCloseToQuaternion.ts` と `toBeCloseToVector3.ts` については、他パッケージにすでにあるファイルと同じはずなので、レビュー不要です。

0b5vr commented on packages/three-vrm-core/src/lookAt/VRMLookAt.ts:
```diff
@@ -239,9 +249,11 @@ export class VRMLookAt {
   public lookAt(position: THREE.Vector3): void {
     // Look at direction in local coordinate
-    const headRotInv = quatInvertCompat(this.getLookAtWorldQuaternion(_quatA));
+    const headRotDiffInv = _quatA
+      .copy(this._restHeadWorldQuaternion)
+      .multiply(quatInvertCompat(this.getLookAtWorldQuaternion(_quatB)));
```
頭のrestからの回転 in world-spaceのinverseです。lookAtを計算するときに、一旦頭のワールド回転のrestからの変化を打ち消すのに使います。
即ち、横や上をむいている場合や、頭を傾げている場合もちゃんとyaw-pitchが取れるようにするもの。

0b5vr commented on packages/three-vrm-core/src/lookAt/VRMLookAtBoneApplier.ts:
```diff
@@ -129,11 +145,20 @@ export class VRMLookAtBoneApplier implements VRMLookAtApplier {
-      // quatB^-1 * quatA * quatB * restQuatLeftEye
-      leftEye.quaternion.copy(_quatB).premultiply(_quatA).premultiply(_quatB.invert()).multiply(this._restQuatLeftEye);
-      leftEyeNormalized!.quaternion.copy(_quatB).premultiply(_quatA).premultiply(_quatB.invert());
+      // _quatB * _quatA * _quatB^-1
+      // where _quatA is LookAt rotation
+      // and _quatB is worldFaceFrontQuat
+      leftEyeNormalized!.quaternion.copy(_quatB).multiply(_quatA).multiply(_quatB.invert());
```
`faceFront` が `(0, 0, -1)` の場合、
一旦 `(0, 1, 0; 0)` (Y軸周りCCW180度) を掛けてから、yaw-pitchから得られたyaw-pitchを掛けて、そのあと `(0, -1, 0; 0)` (Y軸周りCW180度) を掛けて元に戻す

ke456-png replied:
[memo]
> faceFront が (0, 0, -1) の場合

VEC3_POSITIVE_Z と比較

0b5vr committed:
refactor: refactor a quaternion calculation part

0b5vr commented on packages/three-vrm-core/src/lookAt/VRMLookAtBoneApplier.ts:
```diff
@@ -129,11 +144,18 @@ export class VRMLookAtBoneApplier implements VRMLookAtApplier {
+      // and _quatB is worldFaceFrontQuat
+      leftEyeNormalized!.quaternion.copy(_quatB).multiply(_quatA).multiply(_quatB.invert());
+
+      _quatA.copy(this._restLeftEyeParentWorldQuat);
+
+      // _quatA^-1 * leftEyeNormalized.quaternion * _quatA * restQuatLeftEye
+      // where _quatA is restLeftEyeParentWorldQuat
+      leftEye.quaternion.copy(leftEyeNormalized!.quaternion).multiply(_quatA).premultiply(_quatA.invert()).multiply(this._restQuatLeftEye);
```
上でやった `leftEyeNormalized.quaternion` の計算結果をrawにしているだけです。Normalized Human Bonesとやっていることは同じですね。

ke456-png committed:
fix lint error

ke456-png commented on packages/three-vrm-core/src/lookAt/VRMLookAtBoneApplier.ts:
```diff
@@ -90,16 +101,20 @@ export class VRMLookAtBoneApplier implements VRMLookAtApplier {
     // set rest quaternions
     this._restQuatLeftEye = new THREE.Quaternion();
     this._restQuatRightEye = new THREE.Quaternion();
+    this._restLeftEyeParentWorldQuat = new THREE.Quaternion();
+    this._restRightEyeParentWorldQuat = new THREE.Quaternion();
```
[memo]
Normalized -> Rawへ姿勢を適用するため

ke456-png commented on packages/three-vrm-core/src/lookAt/VRMLookAt.ts:
```diff
@@ -105,6 +106,11 @@ export class VRMLookAt {
+  /**
+   * World rotation of the head in its rest pose.
+   */
+  private _restHeadWorldQuaternion: THREE.Quaternion;
```
[memo]
1.0ではtargetからyaw pitchを求めるときと、
yaw pitchで目の姿勢を変更する二つで対応が必要になる

ke456-png reviewed (commented):
LookAtのコードのレビューをして、コメントを返答しました。
ほとんど問題ないと思います。

テストコードのレビューがまだなので、
レビューが済んだらApproveできると思います。

lint errorを修正しました。
[Should] pre-commit hookは動作していますか？

0b5vr commented:
> lint errorを修正しました。 [Should] pre-commit hookは動作していますか？

lint-fixを手動で実行しているということは、やっぱりpre-commit hookは動いていなさそう。。。
@ke456-png さんの環境では動いていますか？

ke456-png commented:
> lint-fixを手動で実行しているということは、やっぱりpre-commit hookは動いていなさそう。。。 @ke456-png さんの環境では動いていますか？

はい、pre-commit hookでlint-stageが実行されていますね

自分も最初は動作していなかっため、.husky/以下の権限を変更しました。
`chmod ug+x .husky/*`

ke456-png commented on packages/three-vrm-core/src/lookAt/tests/VRMLookAtBoneApplier.test.ts:
```diff
@@ -0,0 +1,142 @@
+
+  describe('applyYawPitch', () => {
+    it('makes the raw left eye look toward the specified angle', () => {
+      applier.applyYawPitch(15.0, 15.0);
+
+      const rawLeftEye = humanoid.getRawBoneNode('leftEye')!;
+      const expected = new THREE.Quaternion()
+        .setFromEuler(new THREE.Euler(DEG2RAD * 2.5, DEG2RAD * 2.5, 0.0, 'YXZ'))
+        .multiply(QUAT_Y_CW90);
```
ここの2.5という値の意図がわからなかったので教えてほしいです。

0b5vr replied:
rangeMapの挙動ですね。 `15 / 30 * 5` です。コメント追加しておきます

0b5vr replied:
追加しました @ https://github.com/pixiv/three-vrm/pull/1038/commits/b2b9291ec0b533e2b41182ae3d60c98faa074abc

ke456-png replied:
はい、理解しました。
ありがとうございます。

0b5vr committed:
test: add a comment

Context: https://github.com/pixiv/three-vrm/pull/1038#discussion_r976002657

ke456-png commented on packages/three-vrm-core/src/lookAt/tests/VRMLookAtBoneApplier.test.ts:
```diff
@@ -105,9 +105,12 @@ describe('VRMLookAtBoneApplier', () => {
       applier.applyYawPitch(15.0, 15.0);
 
       const rawLeftEye = humanoid.getRawBoneNode('leftEye')!;
+
+      // saturate( angle / rangeMap.inputMaxValue ) * rangeMap.outputScale = saturate( 15 / 30 ) * 5 = 2.5
```
👍

ke456-png reviewed (approved):
Thanks! 👍
